### ISSUE TITLE
if bound variable inference

### ISSUE THREAD

**Issue by hmc-cs-mdrissi** [microsoft/pyright#1689]
**Describe the bug**
```python

import random

x = random.random() < 0.5

if x:
    y = 5

if x:
    y += 1
```

The second usage of y is reported as potentially unbound. As value of x never changed between the two branches, the second conditional will only be passed if the first one was.

**Expected behavior**
No error.

**VS Code extension or command-line**
pyright command line 1.1.125

**Additional context**
I'm unsure if this is reasonably fixable as I think it boils down to for boolean variables does the system keep track of potential changes. Here ideally a boolean condition being reused without every being modified would indicate its the same value and code only enters the second case if it did the first case. If this is too complex to fix that's fair and I'll just tweak my code.

**Comment by erictraut** [comment]
It would be nice if this were possible, but you're expecting too much from a static type checker. No static type checker that I'm familiar with (in any language) would be able to make that inference. It's already challenging to infer the type of a symbol (say `y`) for arbitrary code flow graphs. It would be exponentially difficult to track types that depend on the values of other symbol values (e.g. `y` is type `int` if the value of `x` is truthy but `unbound` if `x` is not truthy). You would effectively need to execute the program to make those inferences, and then you'd run into the halting problem, etc. It's not feasible.

If you want to gain the benefits of static type checking, you'll need to make some small adjustments to some of your coding patterns. This is an example of where a small workaround is needed. For example, you can preinitialize `y` or conditionally initialize it.

```python
y = 0
if x:
    y = 5
```
or

```python
y = 0 if not x else 5
```
